File: src/info/meodinger/LabelPlusFX/Type/TransFile.java
package info.meodinger.LabelPlusFX.Type;

import info.meodinger.LabelPlusFX.Util.CDialog;
import info.meodinger.LabelPlusFX.Util.CString;

import com.fasterxml.jackson.annotation.JsonAlias;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import java.util.*;

/**
 * @author Meodinger
 * Date: 2021/5/18
 * Location: info.meodinger.LabelPlusFX.Type
 */
public abstract class TransFile<T> {

    public static final int[] DEFAULT_VERSION = new int[] {1, 0};
    public static final String DEFAULT_COMMENT = "Default Comment\nYou can edit me";
    public static final String[] DEFAULT_COMMENT_LIST = {
            DEFAULT_COMMENT, "由 MoeFlow.com 导出", "由MoeTra.com导出"
    };

    private int[] version;
    private String comment;
    @JsonAlias({"group", "groups"})
    private List<T> groupList;
    private Map<String, List<TransLabel>> transMap;

    public int[] getVersion() {
        return version;
    }
    public String getComment() {
        return comment;
    }
    public List<T> getGroupList() {
        return groupList;
    }
    public Map<String, List<TransLabel>> getTransMap() {
        return transMap;
    }

    public void setVersion(int[] version) {
        this.version = version;
    }
    public void setComment(String comment) {
        this.comment = comment;
    }
    public void setGroupList(List<T> groupList) {
        this.groupList = groupList;
    }
    public void setTransMap(Map<String, List<TransLabel>> transMap) {
        this.transMap = transMap;
    }

    @Override
    protected Object clone() throws CloneNotSupportedException {
        return super.clone();
    }

    @Override
    public String toString() {
        String string = "";
        try {
            string = new ObjectMapper().writeValueAsString(this);
        } catch (JsonProcessingException e) {
            CDialog.showException(e);
        }
        return string;
    }

    public static class MeoTransFile extends TransFile<MeoTransFile.Group> {

        @Override
        public MeoTransFile clone() {
            MeoTransFile translation = new MeoTransFile();

            translation.setVersion(this.getVersion().clone());
            translation.setGroupList(new ArrayList<>(this.getGroupList()));
            translation.setComment(this.getComment());
            translation.setTransMap(new HashMap<>(this.getTransMap()));

            return translation;
        }

        public static List<String> getSortedPicList(MeoTransFile file) {
            List<String> trimmed = CString.trimSame(new ArrayList<>(file.getTransMap().keySet()));
            if (trimmed != null && trimmed.size() > 2) {
                boolean canCastToNumberList = true;
                for (int i = 2; i < trimmed.size(); i++) {
                    if (!CString.isDigit(trimmed.get(i))) {
                        canCastToNumberList = false;
                        break;
                    }
                }
                if (canCastToNumberList) {
                    Map<Integer, Integer> map = new HashMap<>();
                    List<Integer> integerList = new ArrayList<>();
                    for (int i = 2; i < trimmed.size(); i++) {
                        int num = Integer.parseInt(trimmed.get(i));
                        integerList.add(num);
                        map.put(num, i);
                    }
                    integerList.sort(Comparator.naturalOrder());

                    int numberLength, complementLength;
                    ArrayList<String> list = new ArrayList<>();
                    for (Integer integer : integerList) {
                        numberLength = CString.lengthOf(integer);
                        complementLength = trimmed.get(map.get(integer)).length() - numberLength;
                        list.add(trimmed.get(0) + CString.repeat('0', complementLength) + integer + trimmed.get(1));
                    }
                    return list;
                }
            }

            Set<String> sorted = new TreeSet<>(Comparator.naturalOrder());
            sorted.addAll(file.getTransMap().keySet());
            return new ArrayList<>(sorted);
        }

        public static String toJsonString(MeoTransFile file) {
            TransFile.MeoTransFile cloned = file.clone();

            List<String> sorted = getSortedPicList(file);
            Map<String, List<TransLabel>> map = new LinkedHashMap<>();
            for (String key : sorted) {
                map.put(key, file.getTransMap().get(key));
            }
            cloned.setTransMap(map);

            ObjectMapper mapper = new ObjectMapper();
            mapper.enable(SerializationFeature.INDENT_OUTPUT);
            String content = "";
            try {
                content = mapper.writeValueAsString(cloned);
            } catch (JsonProcessingException e) {
                e.printStackTrace();
            }

            return content;
        }

        public static class Group {

            public String name;
            public String color;

            // For jackson
            public Group() {}

            public Group(String name, String color) {
                this.name = name;
                this.color = color;
            }
        }
    }

    public static class LPTransFile extends TransFile<String> {

        public final static String PIC_START = ">>>>>>>>[";
        public final static String PIC_END = "]<<<<<<<<";
        public final static String LABEL_START = "----------------[";
        public final static String LABEL_END = "]----------------";
        public final static String PROP_START = "[";
        public final static String PROP_END = "]";
        public final static String SPLIT = ",";
        public final static String SEPARATOR = "-";

    }

}

File: src/main/kotlin/info/meodinger/lpfx/type/TransFile.kt
package info.meodinger.lpfx.type

import info.meodinger.lpfx.util.resource.I18N
import info.meodinger.lpfx.util.resource.get
import info.meodinger.lpfx.util.string.sortByDigit
import info.meodinger.lpfx.util.property.getValue
import info.meodinger.lpfx.util.property.setValue

import com.fasterxml.jackson.annotation.*
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.SerializationFeature
import javafx.beans.property.*
import javafx.collections.FXCollections
import javafx.collections.ObservableList
import javafx.collections.ObservableMap


/**
 * Author: Meodinger
 * Date: 2021/7/29
 * Location: info.meodinger.lpfx.type
 */

/**
 * A MEO Translation file
 */
@JsonIncludeProperties("version", "comment", "groupList", "transMap")
@JsonAutoDetect(getterVisibility = JsonAutoDetect.Visibility.ANY)
open class TransFile @JsonCreator constructor(
    @JsonProperty("version")   version:   IntArray                                    = DEFAULT_VERSION,
    @JsonProperty("comment")   comment:   String                                      = DEFAULT_COMMENT,
    @JsonProperty("groupList") groupList: MutableList<TransGroup>                     = ArrayList(),
    @JsonProperty("transMap")  transMap:  MutableMap<String, MutableList<TransLabel>> = HashMap()
) {

    companion object {

        val DEFAULT_VERSION = intArrayOf(1, 0)

        const val DEFAULT_COMMENT = "使用 LabelPlusFX 导出"
        val DEFAULT_COMMENT_LIST = arrayListOf(
            DEFAULT_COMMENT,
            "Default Comment\nYou can edit me",
            "由 MoeFlow.com 导出",
            "由MoeTra.com导出"
        )

        val DEFAULT_FILE = TransFile()

        object LPTransFile {
            const val PIC_START = ">>>>>>>>["
            const val PIC_END = "]<<<<<<<<"
            const val LABEL_START = "----------------["
            const val LABEL_END = "]----------------"
            const val PROP_START = "["
            const val PROP_END = "]"
            const val SPLIT = ","
            const val SEPARATOR = "-"

            val DEFAULT_COLOR_LIST = listOf(
                "FF0000", "0000FF", "008000",
                "1E90FF", "FFD700", "FF00FF",
                "A0522D", "FF4500", "9400D3"
            )
        }

        // ----- Exception ----- //

        class TransFileException(message: String) : RuntimeException(message) {
            companion object {
                fun pictureNotFound(picName: String) =
                    TransFileException(String.format(I18N["exception.trans_file.picture_not_found.format.s"], picName))

                fun transGroupNameRepeated(groupName: String) =
                    TransFileException(String.format(I18N["exception.trans_file.group_name_repeated.format.s"], groupName))
                fun transGroupIdNegative(groupId: Int) =
                    TransFileException(String.format(I18N["exception.trans_file.group_id_negative.format.i"], groupId))
                fun transGroupIdOutOfBounds(groupId: Int) =
                    TransFileException(String.format(I18N["exception.trans_file.group_id_out_of_bounds.format.i"], groupId))
                fun transGroupNotFound(groupName: String) =
                    TransFileException(String.format(I18N["exception.trans_file.group_not_found.format.s"], groupName))

                fun transLabelIndexRepeated(picName: String, index: Int) =
                    TransFileException(String.format(I18N["exception.trans_file.label_index_repeated.is"], index, picName))
                fun transLabelGroupIdOutOfBounds(groupId: Int) =
                    TransFileException(String.format(I18N["exception.trans_file.label_groupId_out_of_bounds.i"], groupId))
                fun transLabelNotFound(picName: String, index: Int) =
                    TransFileException(String.format(I18N["exception.trans_file.label_not_found.is"], index, picName))
            }
        }

    }

    // ----- Properties ----- //

    val versionProperty: ReadOnlyObjectProperty<IntArray> = SimpleObjectProperty(version)
    val commentProperty: StringProperty = SimpleStringProperty(comment)
    val groupListProperty: ObservableList<TransGroup> = FXCollections.observableList(groupList)
    val transMapProperty: ObservableMap<String, MutableList<TransLabel>> = FXCollections.observableMap(transMap)

    // ----- Accessible Fields ----- //

    val version: IntArray by versionProperty
    var comment: String by commentProperty
    val groupList: List<TransGroup> by groupListProperty
    val transMap: Map<String, List<TransLabel>> by transMapProperty

    val groupCount: Int get() = groupListProperty.size
    val groupNames: List<String> get() = List(groupListProperty.size) { groupListProperty[it].name }
    val groupColors: List<String> get() = List(groupListProperty.size) { groupListProperty[it].colorHex }
    val sortedPicNames: List<String> get() = sortByDigit(transMapProperty.keys.toList())

    // ----- TransGroup ----- //

    fun addTransGroup(transGroup: TransGroup) {
        for (group in groupListProperty)
            if (group.name == transGroup.name)
                throw TransFileException.transGroupNameRepeated(transGroup.name)

        groupListProperty.add(transGroup)
    }
    fun getTransGroup(groupName: String): TransGroup {
        for (group in groupListProperty) if (group.name == groupName) return group

        throw TransFileException.transGroupNotFound(groupName)
    }
    fun getTransGroup(groupId: Int): TransGroup {
        if (groupId < 0) throw TransFileException.transGroupIdNegative(groupId)
        if (groupId >= groupListProperty.size) throw TransFileException.transGroupIdOutOfBounds(groupId)

        return groupListProperty[groupId]
    }
    fun removeTransGroup(groupName: String) {
        var toRemove: TransGroup? = null
        for (group in groupListProperty) if (group.name == groupName) toRemove = group

        if (toRemove != null) groupListProperty.remove(toRemove)
        else throw TransFileException.transGroupNotFound(groupName)
    }
    fun removeTransGroup(groupId: Int) {
        if (groupId < 0) throw TransFileException.transGroupIdNegative(groupId)
        if (groupId >= groupListProperty.size) throw TransFileException.transGroupIdOutOfBounds(groupId)

        groupListProperty.removeAt(groupId)
    }

    fun getGroupIdByName(name: String): Int {
        groupListProperty.forEachIndexed { index, transGroup -> if (transGroup.name == name) return index }

        throw TransFileException.transGroupNotFound(name)
    }
    fun isGroupUnused(groupName: String): Boolean {
        return isGroupUnused(getGroupIdByName(groupName))
    }
    fun isGroupUnused(groupId: Int): Boolean {
        for (key in transMapProperty.keys) for (label in getTransList(key)) {
            if (label.groupId == groupId) return false
        }
        return true
    }

    // ----- TransList (TransMap) ----- //

    open fun addTransList(picName: String) {
        transMapProperty[picName] = ArrayList()
    }
    open fun getTransList(picName: String): MutableList<TransLabel> {
        return transMapProperty[picName] ?: throw TransFileException.pictureNotFound(picName)
    }
    open fun removeTransList(picName: String) {
        if (transMapProperty[picName] != null) transMapProperty.remove(picName)
        else throw TransFileException.pictureNotFound(picName)
    }

    // ----- TransLabel ----- //

    fun addTransLabel(picName: String, transLabel: TransLabel) {
        val list = getTransList(picName)

        val (index, groupId) = transLabel
        if (groupId >= groupListProperty.size) throw TransFileException.transLabelGroupIdOutOfBounds(groupId)
        for (label in list) if (label.index == index) throw TransFileException.transLabelIndexRepeated(picName, index)

        list.add(transLabel)
    }
    fun getTransLabel(picName: String, labelIndex: Int): TransLabel {
        for (label in getTransList(picName)) if (label.index == labelIndex) return label

        throw TransFileException.transLabelNotFound(picName, labelIndex)
    }
    fun removeTransLabel(picName: String, labelIndex: Int) {
        val list = getTransList(picName)
        var toRemove: TransLabel? = null
        for (label in list) if (label.index == labelIndex) toRemove = label

        if (toRemove != null) list.remove(toRemove)
        else throw TransFileException.transLabelNotFound(picName, labelIndex)
    }

    // ----- Other ----- //

    fun toJsonString(): String {
        val mapper = ObjectMapper()

        mapper.enable(SerializationFeature.INDENT_OUTPUT)
        mapper.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY)

        return mapper.writeValueAsString(this.clone())
    }

    fun clone(): TransFile {
        val version = this.version.clone()
        val comment = this.comment
        val groupList = MutableList(this.groupListProperty.size) { this.groupListProperty[it].clone() }
        val transMap = LinkedHashMap<String, MutableList<TransLabel>>().also { map ->
            for (key in sortedPicNames)
                map[key] = MutableList(this.transMapProperty[key]!!.size) {
                    this.transMapProperty[key]!![it].clone()
                }
        }

        return TransFile(version, comment, groupList, transMap)
    }

    override fun toString(): String = ObjectMapper().writeValueAsString(this)

}